src/pudl/transform/ferc1.py

```diff
@@ -249,6 +250,56 @@ def align_row_numbers_dbf(
     return df
 
 
+class SelectDateRangeDurationXbrl(TransformParams):
+    """Parameters for selecting date ranges for report_years in the duration table."""
+
+    date_ranges: list[
+        dict[Literal["report_year", "start_date", "end_date"], int | datetime.date]
+    ] = []
+    """List of date ranges.
+
+    Each element of the list must be a dictionary with three keys: report_year,
+    start_date and end_date. The report_year should be an year represented as an int.
+    This will be the report year that you want to apply this date range to. The start
+    and end date must be dates.
+    """
+
+
+def select_date_range_duration_xbrl(
+    df: pd.DataFrame, params: SelectDateRangeDurationXbrl | None = None
+) -> pd.DataFrame:
+    """Select rows that have specific date ranges for each report_year.
+
+    In the XBRL tables, there are occassionally multiple entries for each report year.
+    Sometimes there are sneaky, sneaky quarterly filings that shouldn't be in the annual
+    FERC1 reporting tables. Sometimes this includes data from the previous year. This
+    function enables us to choose specific start_date and end_date for any specific
+    report_year.
+    """
+    if not params:
+        params = SelectDateRangeDurationXbrl()
+    if params.date_ranges:
+        len_og = len(df)
+        df = df.astype({"start_date": "datetime64", "end_date": "datetime64"})
+        # accumulate the dfs in this list so we can concat everything together later
+        ranges = []
+        for date_range in params.date_ranges:
+            logger.debug(f"selecting date range: {date_range}")
+            ranges.append(
+                df[
+                    (df.report_year == date_range["report_year"])
+                    & (df.start_date.dt.date == date_range["start_date"])
+                    & (df.end_date.dt.date == date_range["end_date"])
+                ]
+            )
+        df = pd.concat(ranges)
```

Comment:
there has got to be a better way to dynamically apply multiple masks... i poked around at this a bit but couldn't find an elegant solution easily.

Reply:
In a boolean sense, the mask you're trying to apply is

```
(date_range_1) | (date_range_2) | (date_range_3)
```
Where each `date_rage` is the three criteria AND-ed together.

But I don't know how you can build up that complicated collection of DF references in a list comprehension.

What's the situation where you imagine needing to parameterize this with several different date ranges?

Reply:
in this setup when we  add another year of data we'll need to add another `date_range`

yeah you can do things like `"|".join(list_of_strings_of_any_length)` but you can't do it with a list of series.

Reply:
so a very simple:
```
def select_current_year_annual_records(df):
    df[
        (df.start_date.dt.date == datetime.date(df.report_year, 1, 1))
        & (df.end_date.dt.date == datetime.date(df.report_year, 12, 31))
    ]
    return df
```

Reply:
I don't love this..... I think it is less explicit. I guess its fine but it definitely does solve the problem of quarterly reports getting slipped in. It also doesn't enable you to keep any end-of-last-year data.

Reply:
It's explicit in that the function name describes what it's doing, and I think it's something we probably want to be doing in a lot of places -- maybe even in every duration table by default. I don't feel like having complex parameters that in practice we only use to describe a single simple operation is great either.  If there's just 2-3 different selections that we do repeatedly, having a readable function name that describes the operation seems okay. If we end up really needing the granular flexibility to do lots of different kinds of selections we could also easily wrap the more flexible function in a stock method with hard coded parameters.

Are there any "make end of last year into start of this year" use cases that are *not* the starting-balance -> ending balance case? I think that case only applies to the instant tables, and will also often be accompanied by reshaping the last year / this year records into columns (like in the plant in service table).

Reply:
okay i tried this on all of the current duration table and in all cases but this one, it results in 100% of the original table. so i made it a default transform.